samples/file-encryptor/host/host.cpp

```diff
@@ -269,104 +277,67 @@ int encrypt_file(
         }
     }
 
-    // The CBC mode for AES assumes that we provide data in blocks of
-    // CIPHER_BLOCK_SIZE bytes. If the file size is not a multiple of
-    // CIPHER_BLOCK_SIZE-byte blocks, PKCS5 Padding was used to make it exactly
-    // a CIPHER_BLOCK_SIZE-byte block
-    if (leftover_bytes)
+    if (encrypt)
     {
-        unsigned char paddingtest[CIPHER_BLOCK_SIZE];
-        unsigned char paddingtest_ciphertext[CIPHER_BLOCK_SIZE];
+        // The CBC mode for AES assumes that we provide data in blocks of
+        // CIPHER_BLOCK_SIZE bytes. PKCS#5 padding was used in this sample. Pad
+        // the whole CIPHER_BLOCK_SIZE block if leftover_bytes is zero. Pad the
+        // (CIPHER_BLOCK_SIZE - leftover_bytes) bytes if leftover_bytes is
+        // non-zero.
+        size_t padded_byte_count = 0;
+        unsigned char plaintext_padding_buf[CIPHER_BLOCK_SIZE];
+        unsigned char ciphertext_padding_buf[CIPHER_BLOCK_SIZE];
+
+        memset(ciphertext_padding_buf, 0, CIPHER_BLOCK_SIZE);
+        memset(plaintext_padding_buf, 0, CIPHER_BLOCK_SIZE);
+
+        if (leftover_bytes == 0)
+            padded_byte_count = CIPHER_BLOCK_SIZE;
+        else
+            padded_byte_count = CIPHER_BLOCK_SIZE - leftover_bytes;
+
         cout << "Host: Working the last block" << endl;
-        cout << "Host: Input file size if not multiples of "
-             << CIPHER_BLOCK_SIZE << "-byte blocks "
-             << "(leftover_bytes = " << leftover_bytes << endl;
+        cout << "Host: padded_byte_count " << padded_byte_count << endl;
+        cout << "Host: leftover_bytes " << leftover_bytes << endl;
+
+        bytes_read = fread(
+            plaintext_padding_buf,
+            sizeof(unsigned char),
+            leftover_bytes,
+            src_file);
+        if (bytes_read != leftover_bytes)
+            goto exit;
+
+        // PKCS5 Padding
```

Comment:
Question: In PKCS5 Padding, you just use the leftover_bytes as the padding value?

Reply:
You use CIPHER_BLOCK_SIZE - leftover_bytes as the padding. For example, if you have 11 bytes left, you need to pad 5 bytes to align it to the 16 byte block size. So, you'd pad it as five 5s. If your message is aligned to the block size, you add an additional block of filled with 16s. This ensures that you can always use the last byte of the padded plaintext to determine the padding.

Reply:
I think it's just for ease of decrypting. You don't need to store the plaintext length and you can use the last byte of the padded plaintext to determine the padding. There are probably other schemes that would do the same thing, but this method was standardized.

Reply:
Akash is right, it's not for more secure reason, it's to make it possible to recognize  where the real end of the ciphertext is.